src/core/ext/xds/xds_bootstrap_grpc.cc

```diff
@@ -36,380 +36,285 @@
-      } else {
-        config_json = it->second;
-      }
-    } else {
-      // "config" is an optional field, so default to an empty JSON object.
-      config_json = Json::Object();
-    }
-    // Try to instantiate the provider.
-    CertificateProviderFactory* factory =
-        CertificateProviderRegistry::LookupCertificateProviderFactory(
-            plugin_name);
-    if (factory == nullptr) {
-      error_list.push_back(GRPC_ERROR_CREATE_FROM_CPP_STRING(
-          absl::StrCat("Unrecognized plugin name: ", plugin_name)));
-    } else if (config_json.has_value()) {
-      grpc_error_handle parse_error = GRPC_ERROR_NONE;
-      RefCountedPtr<CertificateProviderFactory::Config> config =
-          factory->CreateCertificateProviderConfig(*config_json, &parse_error);
-      if (!GRPC_ERROR_IS_NONE(parse_error)) {
-        error_list.push_back(parse_error);
-      } else {
-        certificate_providers_.insert(
-            {instance_name, {std::move(plugin_name), std::move(config)}});
+  void JsonPostLoad(const Json& /*json*/, const JsonArgs& /*args*/,
+                    ErrorList* errors) {
+    // Verify that each authority has the right prefix in the
+    // client_listener_resource_name_template field.
+    {
+      ScopedField field(errors, ".authorities");
+      for (const auto& p : authorities) {
+        const std::string& name = p.first;
+        const Authority& authority = p.second;
+        ScopedField field(
+            errors, absl::StrCat("[\"", name,
+                                 "\"].client_listener_resource_name_template"));
+        std::string expected_prefix = absl::StrCat("xdstp://", name, "/");
+        if (!authority.client_listener_resource_name_template.empty() &&
+            !absl::StartsWith(authority.client_listener_resource_name_template,
+                              expected_prefix)) {
+          errors->AddError(
+              absl::StrCat("field must begin with \"", expected_prefix, "\""));
+        }
       }
     }
   }
-  return GRPC_ERROR_CREATE_FROM_VECTOR_AND_CPP_STRING(
-      absl::StrCat("errors parsing element \"", instance_name, "\""),
-      &error_list);
+};
+
+}  // namespace
+
+//
+// GrpcXdsBootstrap
+//
+
+absl::StatusOr<std::unique_ptr<GrpcXdsBootstrap>> GrpcXdsBootstrap::Create(
+    absl::string_view json_string) {
+  auto json = Json::Parse(json_string);
+  if (!json.ok()) {
+    return absl::InvalidArgumentError(absl::StrCat(
+        "Failed to parse bootstrap JSON string: ", json.status().ToString()));
+  }
+  // Validate JSON.
+  class XdsJsonArgs : public JsonArgs {
+   public:
+    bool IsEnabled(absl::string_view key) const override {
+      if (key == "federation") return XdsFederationEnabled();
+      return true;
+    }
+  };
+  auto bootstrap_json = LoadFromJson<BootstrapJson>(*json, XdsJsonArgs());
+  if (!bootstrap_json.ok()) return bootstrap_json.status();
+  std::vector<XdsServer> servers;
+  for (auto& server : bootstrap_json->servers) {
```

Comment:
I'm certain I'm missing something, but any way we could make this `servers = std::move(bootstrap_json->servers);`?

Reply:
I've changed it to avoid the need for this entirely by making XdsBootstrap a cleaner interface.  The change was a little invasive, but I think it makes things cleaner anyway.